ci/src/main/java/ci/ContinuousIntegrationServer.java

```diff
@@ -176,57 +119,95 @@ public void generateBuildLog(PushStatus ps) {
+        } catch (Exception e) {
             e.printStackTrace();
         }
     }
 
     /**
-     * Parse request data and create a string containing the payload.
-     *
-     * @param request A HttpServletRequest recieved by handler.
-     * @return String Payload of the request.
+     * @param url         Requested URL
+     * @param baseRequest A request instance is created for each HttpConnection accepted by the server
+     * @param response    ServletResponse
+     * @return Boolean
+     * @throws IOException general class of exceptions
      */
-    public String getRequestPayload(HttpServletRequest request) {
-        StringBuilder stringBuilder = new StringBuilder();
-        BufferedReader bufferedReader = null;
+    public boolean buildInUrl(String url, Request baseRequest, HttpServletResponse response) throws IOException {
 
-        try {
-            InputStream inputStream = request.getInputStream();
-            if (inputStream != null) {
-                bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
-                char[] charBuffer = new char[128];
-                int bytesRead = -1;
-                while ((bytesRead = bufferedReader.read(charBuffer)) > 0) {
-                    // Append charvuffer to String builder with given offsets.
-                    stringBuilder.append(charBuffer, 0, bytesRead);
-                }
-            } else {
-                stringBuilder.append("");
-            }
-        } catch (IOException ex) {
-            System.out.println("Error when parsing payload");
-        } finally {
-            try {
-                bufferedReader.close();
-            } catch (IOException ex) {
-                ex.printStackTrace();
+        String subDirectory = "";
+        boolean singleBuild = false;
+        int ifSubDirectoryURL = url.indexOf('/', 8);
+        // true if URL matches http(s)://<host>/*, false if matches http(s)://<host>
+        if (ifSubDirectoryURL != -1) {
+            subDirectory = url.substring(ifSubDirectoryURL);
+            // use regex to see if any specific build is accessed
+            Pattern singleBuildPattern = Pattern.compile("/builds/.");
+            Matcher matcher = singleBuildPattern.matcher(subDirectory);
+            // true if URL matches regex of singleBuildPattern
+            singleBuild = matcher.find();
+        }
+
+        // show list of all stored builds if URL is http(s)://<host>/builds
+        if (subDirectory.equals("/builds")) {
+            response.setContentType("text/html");
+
+            // get all build filenames in buildHistory directory
+            File dir = new File("../buildHistory");
+            String[] buildFileNames = dir.list();
+            // start of simple html file
+            StringBuilder output = new StringBuilder("<!DOCTYPE html>\n<html>\n<body>\n<h1>Build list</h1>");
+
+            // iterate through each build
+            assert buildFileNames != null;
```

Comment:
to run assert, you need to enable the assertion which is off by default (source: https://www.javatpoint.com/assertion-in-java). Maybe it's better to use "if", since I guess it will be pretty messy to enable it with Gradle.

Reply:
No no it is not assert is used here just in case to avoid exceptions specially NullPointerException

Reply:
I know. What I meant was based on the tutorial side that I linked it's written that for these cases you use assert also need to set assertion to on. Here is the part in the tutorial:
![image](https://user-images.githubusercontent.com/89870851/154108067-088126b6-fde8-499b-a49c-57cb8af9e226.png)

I tested assert thing on my own computer so it should be true (see image):
![image](https://user-images.githubusercontent.com/89870851/154107763-94f3825a-68a8-4f21-95b1-df4236398cb5.png)

Reply:
go in the branch issue/62 and try it on your local and see if it will be some fel, because if or assert are actually the same by throwing assertionError

Reply:
wait how do I test it when I can't set the buildFileNames? Do you mean that I write:
            buildFileNames = null;
            assert buildFileNames != null;
I the code when I test?

Is there any different from running our program and my test shown before?

Reply:
I tried locally on the branch issue/62, the assert still does not work (since it does not show any assertionError):
![image](https://user-images.githubusercontent.com/89870851/154122388-75c09750-44d5-4bde-9a9c-9bdc27799758.png)
this is my code:
![image](https://user-images.githubusercontent.com/89870851/154122504-489d3f66-b22c-4309-99f6-c1b423bf8711.png)

Reply:
check last commit,  replaced assertion with if statement @Lantern004